### PR TITLE
Verify that a GenericTypeParamDecl's depth and index are correct

### PR DESCRIPTION
**Pull request by jrose-apple**
(with respect to the containing GenericParamList)

As suggested by @slavapestov in #20091.

### CODE REVIEW

File: lib/AST/ASTVerifier.cpp [hunk 1]
```diff
@@ -2715,6 +2715,37 @@ class Verifier : public ASTWalker {
       verifyCheckedBase(nominal);
     }
 
+    void verifyChecked(GenericTypeParamDecl *GTPD) {
```
Comment (jrose-apple) [inline lib/AST/ASTVerifier.cpp]: Are any of these properties things we can meaningfully check in verifyParsed or verifyCheckedAlways (i.e. "verifyCheckedEvenIfThereAreErrorsAfterParsing")?

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
@slavapestov Review ping

**Review by slavapestov — APPROVED** [review]

File: lib/AST/ASTVerifier.cpp (hunk 1 above)
Comment (jrose-apple) [reply lib/AST/ASTVerifier.cpp]: @slavapestov Question still relevant!

Comment (slavapestov) [reply lib/AST/ASTVerifier.cpp]: verifyCheckedAlways() should be OK, but not verifyParsed() since the depth isn't set yet.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test Linux

**Comment by jrose-apple** [comment]
Looks like this has [problems in module merging](https://ci.swift.org/view/Source%20Compatibility/job/swift-master-source-compat-suite-debug/lastFailedBuild/artifact/swift-source-compat-suite/). I'll get back to it soon, but reverting for now.

**Pull request merged**